File: java/Code19.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Code19 {
  public static void main(String[] args) throws IOException {
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(in.readLine());
    int N = Integer.parseInt(st.nextToken());// 데이터 개수
    int K = Integer.parseInt(st.nextToken());// K번째 수
    st = new StringTokenizer(in.readLine());
    int[] A = new int[N];
    for (int i = 0; i < N; i++) {
      A[i] = Integer.parseInt(st.nextToken());
    }
    quickSort(A, 0, N - 1,  K - 1); // 배열, start , end, K번째 수라해도 배열에서는 k-1과 같다. 2번째 수 는 k[1]
    System.out.println(A[K - 1]);
  }
  public static void quickSort(int[] A, int S, int E, int K) {
      if(S<E){
      int pivot = partition(A, S, E);
      if (pivot == K) { //K번째 수가 pivot이면 더이상 구할 필요 없음
        return;
      }
      else if(K < pivot) {  //K가 pivot보다 작으면 왼쪽만 정렬
        quickSort(A, S, pivot - 1, K); //재귀 함수 호출
      }
      else if(K > pivot){  //K가 pivot보다 크면 왼쪽만 정렬
        quickSort(A, pivot+1, E, K); //start 위치  pivot+1  ~ end
      }
          }
  }

  private static int partition(int[] A, int S, int E){
    if(S+1==E) { //앞의 애가 뒤에 애보다 1밖에 차이 안날때 그냥 교환 해준다. 
      if(A[S]>A[E])swap(A,S,E);
      return E;
    }
    int M = (S + E) / 2;   //중앙값을 피벗으로 설정
    swap(A, S, M); // 중앙값을 1번째 요소로 이동하기 A[0] =2 , A[2] =4
    int pivot = A[S];
    int i = S+1, j = E;
    
    while (i <= j) {
        while (j >= S + 1 && pivot < A[j] ){   //피벗보다 작은 수가 나올때까지 j--
            j--;    
        }
        while (i <= E && pivot > A[i]){  //피벗보다 큰 수가 나올 떄까지 i++
               i ++;  
        }
        if (i <= j) {
            swap (A, i++, j--);  // 찾은 i와 j를 교환하기
        }else {
        	break;
        }
    }
    // i == j 피벗의 값을 양쪽으로 분리한 가운데에 오도록 설정하기
    A[S] = A[j];
    A[j] = pivot;
    return j; //피벗의 위치를 반환
}
  
  public static void swap(int[] A, int i, int j) { // swap 기능을 swap함수로 만들어준다.
    int temp = A[i];
    A[i] = A[j];
    A[j] = temp;
  }
}


//중앙값을 피벗으로 설정했지만 pivot은 실질적으로 데이터에서 어떤 값을 설정해도 상관없다고 한다. 
//피벗 값을 중심으로 피벗보다 작은 데이터와 피벗보다 큰 데이터로 나누어주는 로직만 잘 구현이 되면 된다.

File: python/Code19.py
import sys

input = sys.stdin.readline
N, K = map(int, input().split())
A = list(map(int, input().split()))


def quickSort(S, E, K):
    global A
    if S < E:
        pivot = partition(S, E)
        if pivot == K:  # K번째 수가 pivot이면 더이상 구할 필요 없음
            return
        elif K < pivot:  # K가 pivot보다 작으면 왼쪽 그룹만 정렬
            quickSort(S, pivot - 1, K)
        else:  # K가 pivot보다 크면 오른쪽 그룹만 정렬
            quickSort(pivot + 1, E, K)


def swap(i, j):
    global A
    temp = A[i]
    A[i] = A[j]
    A[j] = temp


def partition(S, E):
    global A

    if S + 1 == E:
        if A[S] > A[E]:
            swap(S, E)
        return E

    M = (S + E) // 2
    swap(S, M)
    pivot = A[S]
    i = S + 1
    j = E
    while i <= j:
        while j >= S and pivot < A[j]:
            j = j - 1
        while  i <= E and pivot > A[i] :
            i = i + 1
        if i < j:
            swap(i, j)
            i = i + 1
            j = j - 1
        else:
            break
    # i == j 피벗의 값을 양쪽으로 분리한 가운데에 오도록 설정하기
    A[S] = A[j]
    A[j] = pivot
    return j


quickSort(0, N - 1, K - 1)
print(A[K - 1])

# global A 라는 것은 A라는 전역변수를 사용하는 문법이다. 수 안에서 global을 붙여도 되고, 함수 밖에서 global을 붙여도 된다. 하지만 함수 밖에서 선언하였다면 함수 안에서 사용을 명시해주어야 한다.
# 전체 코드에서 체크해야 하는 값이 있을 때 전역변수를 사용하면 편리하다.
# 특히 함수가 있는 코드를 실행하거나 수정할 때, 에러가 발생하거나 값이 안 맞으면 확인하기가 곤란하다.
#  이럴 때 함수 안의 변수를 전역변수로 선언하면, 코드를 실행하며 함수 안의 값들을 체크할 수 있기 때문에 유용하다. 프로그램을 디버깅할 때 편리한 방법인 듯 하다.
# 여기서는 변수 입력받는 변수하나를 줄여주는 역할을 했다. 원래라면 quicksort 안에 들어가는 매개변수는 리스트 A를 추가한 4개가 됬어야 한다. 